pkg/diag/colors/colors.go

```diff
@@ -69,26 +64,117 @@ func TrimPartialCommand(s string) string {
+	case BrightCyan: // command("fg 14")
+		writeCodes(buf, "38", "5", "14")
+	case RedBackground: // command("bg 1")
+		writeCodes(buf, "48", "5", "1")
+	case GreenBackground: // command("bg 2")
+		writeCodes(buf, "48", "5", "2")
+	case YellowBackground: // command("bg 3")
+		writeCodes(buf, "48", "5", "3")
+	case BlueBackground: // command("bg 4")
+		writeCodes(buf, "48", "5", "4")
+	case Black: // command("fg 0") // Only use with background colors.
+		writeCodes(buf, "38", "5", "0")
+	}
 }
 
-func colorizeText(s string) string {
-	style, err := loreley.Compile(s, nil /*extensions*/)
-	contract.Assertf(err == nil, "Expected no errors during string colorization; str=%v, err=%v", s, err)
-
-	// Note: we only get into this codepath if we determined colors should be on.  This was either
-	// because we were in color=auto mode and our env detection determined colors were reasonable to
-	// have, or because we're in color=always mode.
-	//
-	// However, Loreley does its own tty detection and disables colors if there is no tty.  We don't
-	// want that behavior if we've determined colors should be on.  So we explicitly override their
-	// automatic detection.
-	style.NoColors = false
-	result, err := style.ExecuteToString(nil /*data*/)
-	contract.Assertf(err == nil, "Expected no errors during string colorization; str=%v, err=%v", s, err)
-
-	return result
+func colorizeText(s string, c Colorization, maxLen int) string {
+	var buf bytes.Buffer
+
+	textLen := 0
+	for input := s; len(input) > 0; {
+		// Do we have another directive to process?
+		nextDirectiveStart := strings.Index(input, colorLeft)
+		if nextDirectiveStart == -1 {
+			// If there are no more directives and we still have the entire original string, return it as-is: there
+			// must not have been any directives.
+			if len(input) == len(s) {
+				if maxLen >= 0 && len(input) > maxLen {
+					return input[:maxLen]
+				}
+				return input
+			}
+
+			// Otherwise, write the remaining input into the buffer and terminate.
+			_, err := buf.WriteString(input)
+			contract.IgnoreError(err)
+			break
+		}
+		if buf.Cap() < len(input) {
+			buf.Grow(len(input))
+		}
+
+		// Copy the text up to but not including the delimiter into the buffer.
+		text := input[:nextDirectiveStart]
+		if maxLen >= 0 && textLen+len(text) > maxLen {
+			_, err := buf.WriteString(text[:maxLen])
+			contract.IgnoreError(err)
+			writeDirective(&buf, c, Reset)
+			break
+		}
+		_, err := buf.WriteString(input[:nextDirectiveStart])
+		contract.IgnoreError(err)
+		input = input[nextDirectiveStart+len(colorLeft):]
+
+		// If we have a start delimiter but no end delimiter, terminate. The partial command will not be present in the
+		// output.
+		nextDirectiveEnd := strings.Index(input, colorRight)
+		if nextDirectiveEnd == -1 {
+			break
+		}
+
+		directive := command(input[:nextDirectiveEnd])
+		writeDirective(&buf, c, directive)
+		input = input[nextDirectiveEnd+len(colorRight):]
+	}
+
+	return buf.String()
```

Comment:
Is this all brand new code, or is it derived from parts of lorelay or another library?  Either way - is this something we should extract out and test more completely?

Reply:
This is brand new, though it is compatible with the way in which we were using Loreley. We should add unit tests, yes.